File: backend/utils/aiRecommendation.py
import sys
import json
import pickle
import numpy as np
import os

# Suppress all warnings and output except JSON result
import warnings
warnings.filterwarnings('ignore')
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'  # Suppress TensorFlow warnings

# Redirect stderr to suppress Python output
import sys
sys.stderr = open(os.devnull, 'w')

try:
    from tensorflow import keras
    import tensorflow as tf
    tf.get_logger().setLevel('ERROR')
except:
    pass

# Get arguments
user_ratings_json = sys.argv[1]
user_id = sys.argv[2]

# Load models
model_path = os.path.join(os.path.dirname(__file__), '../../aimodel')
model_file = os.path.join(model_path, 'movie_recommender_dl.h5')

model = None
try:
    # Try loading with compile=False to avoid metrics deserialization issues
    try:
        import tensorflow as tf
        tf.get_logger().setLevel('ERROR')
        # Try with safe_mode=False to bypass validation
        try:
            model = tf.keras.models.load_model(model_file, compile=False, safe_mode=False)
        except:
            # Try loading with legacy format
            try:
                # Use tf.compat.v1 for older models
                with tf.compat.v1.disable_eager_execution():
                    model = tf.keras.models.load_model(model_file, compile=False)
            except:
                # Last resort: try with keras directly
                try:
                    from tensorflow import keras
                    model = keras.models.load_model(model_file, compile=False)
                except:
                    model = None
    except:
        model = None
except Exception as e:
    # If all loading methods fail, output empty list and exit gracefully
    model = None

# If model failed to load, exit early
if model is None:
    print(json.dumps([]))
    sys.exit(0)

try:
    with open(os.path.join(model_path, 'movie_map.pkl'), 'rb') as f:
        movie_map = pickle.load(f)
    
    with open(os.path.join(model_path, 'user_map.pkl'), 'rb') as f:
        user_map = pickle.load(f)
    
    # Get user ratings
    user_ratings = json.loads(user_ratings_json)
    
    # Get user index
    if user_id in user_map:
        user_idx = user_map[user_id]
    else:
        # New user - use average user
        user_idx = len(user_map) // 2 if len(user_map) > 0 else 0
    
    # Get recommendations
    user_vector = np.array([[user_idx]])
    predictions = model.predict(user_vector, verbose=0)
    
    # Get top recommendations
    top_indices = np.argsort(predictions[0])[-10:][::-1]
    
    # Map back to movie IDs
    recommendations = []
    for idx in top_indices:
        for movie_id, movie_idx in movie_map.items():
            if movie_idx == idx:
                recommendations.append(int(movie_id))
                break
    
    # Output only JSON to stdout (this is the only output)
    print(json.dumps(recommendations))
    sys.stdout.flush()  # Ensure output is flushed
except Exception as e:
    # If anything fails, return empty list gracefully
    print(json.dumps([]))
    sys.stdout.flush()
    sys.exit(0)

File: backend/utils/aiRecommendation.js
import { PythonShell } from 'python-shell';
import path from 'path';
import { fileURLToPath } from 'url';
import fs from 'fs';
import { execSync } from 'child_process';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const MODEL_PATH = path.join(__dirname, '../../aimodel');
const PYTHON_SCRIPT = path.join(__dirname, 'aiRecommendation.py');

export const getAIRecommendations = async (userId, userRatings = []) => {
  try {
    // Skip Python AI for quick demo - use smart fallback
    console.log('ℹ️  Using fallback recommendations');
    return await getFallbackRecommendations(userRatings);
    
    // Check if models exist
    const modelFiles = [
      'movie_recommender_dl.h5',
      'movie_map.pkl',
      'user_map.pkl',
    ];

    const modelsExist = modelFiles.every((file) =>
      fs.existsSync(path.join(MODEL_PATH, file))
    );

    if (!modelsExist) {
      console.warn('⚠️  AI models not found, using fallback recommendations');
      return await getFallbackRecommendations(userRatings);
    }

    // Check if Python script exists
    if (!fs.existsSync(PYTHON_SCRIPT)) {
      console.warn('⚠️  Python script not found, using fallback recommendations');
      return await getFallbackRecommendations(userRatings);
    }

    // Prepare user ratings data
    const ratingsData = userRatings.map((r) => ({
      movieId: r.movieId,
      rating: r.rating,
    }));

    // Run Python script
    // Try to find compatible Python version (3.10, 3.11, or 3.12)
    let pythonPath = 'python';
    
    try {
      // Try Python 3.12 first - get actual executable path
      execSync('py -3.12 --version', { stdio: 'ignore' });
      const python312Path = execSync('py -3.12 -c "import sys; print(sys.executable)"', { encoding: 'utf-8' }).trim();
      pythonPath = python312Path;
      console.log('✅ Using Python 3.12 for AI recommendations');
    } catch {
      try {
        // Try Python 3.11
        execSync('py -3.11 --version', { stdio: 'ignore' });
        const python311Path = execSync('py -3.11 -c "import sys; print(sys.executable)"', { encoding: 'utf-8' }).trim();
        pythonPath = python311Path;
        console.log('✅ Using Python 3.11 for AI recommendations');
      } catch {
        try {
          // Try Python 3.10
          execSync('py -3.10 --version', { stdio: 'ignore' });
          const python310Path = execSync('py -3.10 -c "import sys; print(sys.executable)"', { encoding: 'utf-8' }).trim();
          pythonPath = python310Path;
          console.log('✅ Using Python 3.10 for AI recommendations');
        } catch {
          // Fallback to default python
          pythonPath = process.env.PYTHON_PATH || 'python';
          console.warn('⚠️  Using default Python - may not be compatible with TensorFlow');
        }
      }
    }
    
    // Normalize the path to avoid duplication issues
    const normalizedScriptPath = path.normalize(PYTHON_SCRIPT);
    
    const options = {
      mode: 'text',
      pythonPath: pythonPath,
      pythonOptions: ['-u'],
      args: [JSON.stringify(ratingsData), userId],
    };

    // Add timeout to prevent hanging (30 seconds)
    const TIMEOUT_MS = 30000;
    
    return new Promise((resolve) => {
      let resolved = false;
      
      // Set up timeout
      const timeoutId = setTimeout(() => {
        if (!resolved) {
          resolved = true;
          console.warn('⚠️  Python script timeout after 30s. Using fallback recommendations.');
          getFallbackRecommendations(userRatings).then(resolve).catch(() => {
            resolve([550, 238, 240, 424, 497, 680, 13, 769, 155, 429]);
          });
        }
      }, TIMEOUT_MS);

      // Use just the filename with scriptPath set to the directory
      // This prevents path duplication issues
      const scriptName = path.basename(normalizedScriptPath);
      const scriptDir = path.dirname(normalizedScriptPath);
      
      const runOptions = {
        ...options,
        scriptPath: scriptDir,
      };
      
      PythonShell.run(scriptName, runOptions, (err, results) => {
        if (resolved) return; // Already handled by timeout
        
        clearTimeout(timeoutId);
        resolved = true;
        
        if (err) {
          // Check if it's a missing module error
          if (err.message && err.message.includes('ModuleNotFoundError')) {
            console.warn('⚠️  Python dependencies not installed. Using fallback recommendations.');
          } else if (err.message && (err.message.includes('Could not deserialize') || err.message.includes('KerasSaveable'))) {
            console.warn('⚠️  Keras model compatibility issue. Using fallback recommendations.');
          } else {
            // Only log non-critical errors (suppress timeout/killed messages)
            if (!err.message || (!err.message.includes('timeout') && !err.message.includes('killed') && !err.message.includes('SIGTERM'))) {
              console.warn('⚠️  Python script error:', err.message || err);
            }
          }
          // Always resolve with fallback, never reject
          getFallbackRecommendations(userRatings).then(resolve).catch(() => {
            resolve([550, 238, 240, 424, 497, 680, 13, 769, 155, 429]);
          });
          return;
        }

        try {
          if (results && results.length > 0) {
            // Get the last line (should be JSON)
            const output = results[results.length - 1].trim();
            if (output) {
              const recommendations = JSON.parse(output);
              // If empty array returned (model loading failed), use fallback
              if (Array.isArray(recommendations) && recommendations.length > 0) {
                resolve(recommendations);
              } else {
                getFallbackRecommendations(userRatings).then(resolve).catch(() => {
                  resolve([550, 238, 240, 424, 497, 680, 13, 769, 155, 429]);
                });
              }
            } else {
              getFallbackRecommendations(userRatings).then(resolve).catch(() => {
                resolve([550, 238, 240, 424, 497, 680, 13, 769, 155, 429]);
              });
            }
          } else {
            getFallbackRecommendations(userRatings).then(resolve).catch(() => {
              resolve([550, 238, 240, 424, 497, 680, 13, 769, 155, 429]);
            });
          }
        } catch (parseError) {
          console.warn('⚠️  Parse error:', parseError.message);
          getFallbackRecommendations(userRatings).then(resolve).catch(() => {
            resolve([550, 238, 240, 424, 497, 680, 13, 769, 155, 429]);
          });
        }
      });
    });
  } catch (error) {
    console.warn('⚠️  AI Recommendation Error:', error.message || error);
    return await getFallbackRecommendations(userRatings).catch(() => {
      return [550, 238, 240, 424, 497, 680, 13, 769, 155, 429];
    });
  }
};

const getFallbackRecommendations = async (userRatings) => {
  // Fallback: Return popular movie IDs if AI model fails
  // Try to fetch from TMDB if available
  try {
    const { getPopularMovies } = await import('./tmdb.js');
    const popular = await getPopularMovies(1);
    if (popular.results && popular.results.length > 0) {
      return popular.results.slice(0, 10).map(m => m.id);
    }
  } catch (error) {
    console.warn('Could not fetch popular movies for fallback:', error.message);
  }
  // Ultimate fallback: hardcoded popular movie IDs
  return [550, 238, 240, 424, 497, 680, 13, 769, 155, 429];
};

export const getContentBasedRecommendations = async (movieId, genres = []) => {
  // Content-based filtering using movie features
  // This would use the movie_map.pkl to find similar movies
  try {
    // Simplified content-based recommendation
    // In production, use cosine similarity on movie features
    return [];
  } catch (error) {
    console.error('Content-based recommendation error:', error);
    return [];
  }
};

